src/cpp/src/sampler.cpp

```diff
@@ -560,15 +561,89 @@ std::vector<int64_t> Sampler::_try_finish_generation(SequenceGroup::Ptr & sequen
+                        Sequence::Ptr running_sequence,
+                        LogitProcessor& logit_processor,
+                        bool is_extend_sequence,
+                        bool is_update_len_logit_processor) {
+    logit_processor.register_new_generated_token(sampled_token_id.m_index);
+    size_t generated_len = logit_processor.get_generated_len();
+    if (is_extend_sequence) {
+        running_sequence->append_token(sampled_token_id.m_index, sampled_token_id.m_log_prob);
+    } else {
+        // just update the token log prob in case of successfully validated token
+        OPENVINO_ASSERT(generated_len < running_sequence->get_generated_len());
+        running_sequence->update_generated_log_prob(generated_len, sampled_token_id.m_log_prob);
+    }
+    // increment seq len only for one sequence in sequence group to sync them
+    if (is_update_len_logit_processor) {
+        logit_processor.update_generated_len(++generated_len);
+    }
+};
+
+std::list<uint64_t>
+create_n_forked_sequences(SequenceGroup::Ptr sequence_group,
+                          LogitProcessor& logit_processor,
+                          const std::vector<Token>& sampled_tokens) {
+    const auto& running_sequences = sequence_group->get_running_sequences();
+    OPENVINO_ASSERT(running_sequences.size() == 1);
+    Sequence::Ptr sequence_to_fork = running_sequences.front();
+    std::list<uint64_t> forked_seq_ids;
+    for (size_t i = 1; i < sampled_tokens.size(); ++i) {
+        const auto forked_sequence = sequence_group->fork_sequence(sequence_to_fork);
+        const auto forked_seq_id = forked_sequence->get_id();
+        forked_seq_ids.push_back(forked_seq_id);
+        register_new_token(sampled_tokens[i], forked_sequence, logit_processor, true, false);
+    }
+    return forked_seq_ids;
+}
+
+bool
+is_continue_to_sample_tokens(Sequence::Ptr running_sequence,
+                             size_t token_idx,
+                             size_t max_gen_len,
+                             size_t& decrease_context_len_per_seq_group) {
+    if (max_gen_len == 0) {
+        running_sequence->remove_last_tokens(token_idx);
+        decrease_context_len_per_seq_group = std::max(decrease_context_len_per_seq_group, token_idx);
+        return false;
+    }
+    return true;
+}
+
+bool
+validate_candidate(Sequence::Ptr running_sequence,
+                   size_t& token_idx,
+                   Token& sampled_token,
+                   bool& is_extend_sequence,
+                   size_t& decrease_context_len_per_seq_group) {
+    if (token_idx > 0) {
+        const auto& generated_tokens = running_sequence->get_generated_ids();
+        auto it = generated_tokens.rbegin();
+        std::advance(it, token_idx - 1);
+        // to validate candidates from assisting model and remove incorrect ones from generated sequence
+        if (*it != sampled_token.m_index) {
+            running_sequence->remove_last_tokens(token_idx);
```

Comment:
do you think we can also update `context_len` inside this method instead of calling `update_processed_tokens_num` separately ?

Reply:
It is done  per request by 
```
        // decrease sequence_group context in case of candidates generated by draft_model were not accepted by main_model
        if (decrease_context_len_per_seq_group) {
            const auto num_processed_tokens = sequence_group->get_num_processed_tokens();
            OPENVINO_ASSERT(num_processed_tokens >= decrease_context_len_per_seq_group);
            OPENVINO_ASSERT(sequence_group->get_context_len() >= decrease_context_len_per_seq_group);
            sequence_group->update_processed_tokens_num(num_processed_tokens - decrease_context_len_per_seq_group);
        }
```